fendermint/vm/topdown/src/convert.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! Handles the type conversion to ethers contract types
+
+use crate::IPCParentFinality;
+use anyhow::anyhow;
+use ethers::abi::Token;
+use ethers::types::Bytes;
+use ethers::types::U256;
+use fendermint_vm_ipc_actors::{gateway_getter_facet, gateway_router_facet};
+use fvm_shared::address::{Address, Payload};
+use fvm_shared::clock::ChainEpoch;
+use ipc_agent_sdk::message::ipc::ValidatorSet;
+use std::marker::PhantomData;
+use std::str::FromStr;
+
+const COMMIT_PARENT_FINALITY_FUNC_NAME: &str = "commitParentFinality";
+const GET_PARENT_FINALITY_FUNC_NAME: &str = "getParentFinality";
+
+impl From<IPCParentFinality> for gateway_router_facet::ParentFinality {
+    fn from(value: IPCParentFinality) -> Self {
+        let mut array = [0u8; 32];
+        array.copy_from_slice(&value.block_hash[0..32]);
+
+        Self {
+            height: U256::from(value.height),
+            block_hash: array,
+        }
+    }
+}
+
+impl From<gateway_getter_facet::ParentFinality> for IPCParentFinality {
+    fn from(value: gateway_getter_facet::ParentFinality) -> Self {
+        IPCParentFinality {
+            height: value.height.as_u64(),
+            block_hash: value.block_hash.to_vec(),
+        }
+    }
+}
+
+/// Converts a Rust type FVM address into its underlying payload
+/// so it can be represented internally in a Solidity contract.
+fn addr_payload_to_bytes(payload: Payload) -> Bytes {
+    match payload {
+        Payload::Secp256k1(v) => ethers::types::Bytes::from(v),
+        Payload::Delegated(d) => {
+            let addr = d.subaddress();
+            let b = ethers::abi::encode(&[Token::Tuple(vec![
+                Token::Uint(U256::from(d.namespace())),
+                Token::Uint(U256::from(addr.len())),
+                Token::Bytes(addr.to_vec()),
+            ])]);
+            ethers::types::Bytes::from(b)
+        }
+        _ => unimplemented!(),
+    }
+}
+
+fn convert_addr(addr: Address) -> gateway_router_facet::FvmAddress {
+    gateway_router_facet::FvmAddress {
+        addr_type: addr.protocol() as u8,
+        payload: addr_payload_to_bytes(addr.into_payload()),
+    }
+}
+
+/// The decode function call return value in solidity
+pub struct DecodeFunctionReturn<T> {
+    _phantom: PhantomData<T>,
+}
+
+/// The equivalent of abi.encodeWithSignature in solidity
+pub struct EncodeWithSignature<T> {
+    _phantom: PhantomData<T>,
+}
```

Comment:
This might make sense to someone familiar with Solidity, but I find the appearance of "Signature" confusing.

```
    let params = EncodeWithSignature::<
        gateway_router_facet::CommitParentFinalityCall,
    >::encode(finality, validator_set)?;
```

given the complexity of these facets, I'd think this crate should try to hide it and make it simpler to use, rather than make it required knowledge for something like the ChainMessageInterpreter to come up with the right combination of a confusing trait name generic over an ever shifting land of facets and their method groupings. 

A simple method like `encode_commit_parent_finality_call` would provide less flare but be way easier to discover. It doesn't look like there's much commonality between these traits anyway apart from being implemented on calls, but like I said since they are very difficult to discover, you have to know which one lives in which Solidity contract, I would try to hide them at all costs.